content/asm_4.md

```diff
@@ -1,194 +1,312 @@
+;; Calculate the length of the input string and prepare to reverse it.
+reverseStringAndPrint:
+        ;; Compare the first element in the given string with the NUL terminator (end of the string).
+        cmp byte [rsi], 0
+        ;; If we reached the end of the input string, reverse it.
+        je reverseString
+        ;; Load byte from the rsi to al register and move pointer to the next character in the string.
+        lodsb
+        ;; Save the character of the input string on the stack.
+        push rax
+        ;; Increase the counter that stores the length of our input string.
+        inc rcx
+        ;; Continue to go over the input string if we did not reach end of if.
+        jmp reverseStringAndPrint
 ```
 
-and exit from the our program:
+In the beginning of this procedure we compare the value pointed by the `rsi` register with `0`. Here `0` means is the end of the input string as each string is [NUL terminated](https://en.wikipedia.org/wiki/Null-terminated_string). If we reached the end of the input string we jump to the `reverseString` label that will reverse our string. Otherwise we take character by character from the input string with the `loadsb` instruction and store them on the stack. It is very convenient to store the string's characters on the stack. Since the stack has [LIFO](https://en.wikipedia.org/wiki/Stack_(abstract_data_type)) access pattern, we will pop the characters from the stack in reversed order - from the end to the beginning. That actually allows us to reverse the string. This procedure is [recursive](https://en.wikipedia.org/wiki/Recursion) and will be called again and again while we will not reach the end of the string.
+
+After we reached the end of the string we can reverse it. We will do it using the following code:
 
 ```assembly
-exit:
-		mov rax, SYS_EXIT
-		mov rdi, EXIT_CODE
-		syscall
+;; Reverse the string and store it in the output buffer.
+reverseString:
+        ;; Check the counter that stores the length of the string.
+        cmp rcx, 0
+        ;; If it is equal to `0`, print the reverse string.
+        je printResult
+        ;; Pop the character from the stack.
+        pop rax
+        ;; Put the character to the output buffer.
+        mov [rdi], rax
+        ;; Move the pointer to the next character in the output buffer.
+        inc rdi
+        ;; Decrease the counter of the length of the string.
+        dec rcx
+        ;; Move to the next character while we did not reach the end of the string.
+        jmp reverseString
 ```
 
-That's all, now we can compile our program with:
+In the first two lines of the code we check the counter that stores the length of our string and if it is equal to `0`, it means that we finished to reverse the string and time to push it with the `printResult` procedure. Otherwise, we pop the character from the stack to the `rax` register and put it to the address pointed by the `rdi`. During these operations we do two additional procedures: 
```

Comment:
```suggestion
In the first two lines of the code, we check the counter that stores the length of our string. If it is equal to `0`, the string reversal is finished, and we can push it with the `printResult` procedure. Otherwise, we pop the character from the stack to the `rax` register and put it to the address pointed by the `rdi` register. During these operations, we call two additional procedures to: 
```

Reply:
which procedures?